### PR TITLE
Docs: Document tutorial for running K6 Extension (Prometheus Output) with Job Executor Service

### LINKED ISSUE #19: [DOCS] Document Keptn + Keptn (+ prom) so far
Please document what we have working so far at the midway point of GSoC. I suggest doing so on the readme (or introducing a `docs` folder with multiple files if you want).

I suggest doing these docs **before** any more coding is completed. LEt's give users a way to follow our journey and use what we have so far (as it is already extremely valuable).

## Definition of Done
- [ ] PR Raised for: 1) documentation on Keptn + k6 running with Job Executor service
- [ ] An "extended" section of the docs above with instructions (should the user want to export metrics) for how to build and use k6 + Prometheus exporter
- [ ] Above PR is reviewed, changed if appropriate and approved
- [ ] Docs are merged
- [ ] Blog post is written (@agardnerIT happy to help with this)
- [ ] Blog post submitted to Keptn community for publishing on Medium (either via @jainammm's personal site if available or on main Keptn site)

### PR DESCRIPTION
**Pull request by jainammm**
- [x] Add Readme
- [x] Test the flow

Closes #19

### CODE REVIEW

**Review by agardnerIT — COMMENTED** [review]
Looks really great. A few tiny suggestions:

1) Link to Youtube with a time jump. Use the `t=` parameters to skip any intros or waffle take users right to the action `t` parameter is specified in seconds from 0 (start of the video). eg: `https://youtu.be/9VI6GMUfI3g?t=393`

2) I'd advise linking to installation instructions (like JES and Prometheus) rather than including htem here. If the third party service (JES or Prom) updates their installation instructions, you don't need to keep your copy up to date. Less maintenance and headache for you. Every instructions or code snippet here today is tomorrow's "help me" message in slack :)

Other than that, it's fantastic!

**Review by ppcano — COMMENTED** [review]
Shared via Slack.

1 - It is good -clear and direct. I didn’t quickly understand how all the pieces were connected. Note that I don’t know how Keptn works,  any diagram or any additional context about the role of each component would be useful. 

1.1 For example, I think `microserviceA` is not a descriptive name. I suggest renaming it to `k6-performance-test` or something along the line.

2 - I think the titles of the tutorials could be more descriptive.
Running k6 tests in Keptn workflows?

3 - I suggest linking both tutorials from the project Readme

**Comment by jainammm** [comment]
@agardnerIT, made changes as you suggested

@ppcano:
1. added workflow wrt Keptn... the name `microserviceA` is the microservice which will go through these tasks of `test`, `evaluation`, so that should be appropriate
2. Changed the titles
3. Will add in another PR after merging this

PS: @agardnerIT @ppcano please also review the tutorial of quality gates for part 3 too... have added in this PR only

**Review by ppcano — APPROVED** [review]

**Review by agardnerIT — CHANGES_REQUESTED** [review]
Looks like a lot here but I promise they're all small things. Excellent tutorial and I believe the basis of some good conference talks :)

# docs/k6-jes-example/README.md

- You mention the `add-resource` CLi command, mention that users can skip this step and instead upload directly to Git if they choose. They need to upload into the `{branch-name-which-matches-the-stage}/{keptn-service-name}/{resourceUri path}`
For example `production` branch, folder `microserviceA/job/config.yaml`

- You mention that a sequence can be done from the CLI or bridge. Please also mention that sequences can be triggered via the Keptn API.

- "The file mentioned would be accessible from /keptn/<resource-uri>".
This could be worded more clearly to mention that inside the container, files paths are placed in the `/keptn` folder. So `/files/foo.txt` in Git becomes `/keptn/files/foo.txt` inside the container.

- Note: This review will not comment on anything related to k6 instructions as I'm not the expert.

- Prometheus: you mention enabling remote write. I'm not a Prom expert but I know when I made [this](https://github.com/agardnerIT/jes-metric-exporter/blob/main/prometheus.md) I didn't have to do any of that?

# docs/k6-prometheus-quality-gate-example/README.md

- Important to note here that other SLI providers work equally well, not just Prometheus but Dynatrace, Datadog or anything else they bring.
Perhaps add wording to mention that you have chosen to use Prom but they can use whatever they like. It's also an opportunity to crosslink to the Custom SLI Provider page: https://keptn.sh/docs/0.19.x/integrations/sli_provider

"Lighthouse Service will be responsible for valuating test results. Whenever an event of the type sh.keptn.event.evaluation.triggered is received, it will take first determine the SLI (Service Level Indicator) data source (e.g., Prometheus or Dynatrace), as well as the required SLIs for the project. Afterwards, it will send out an event of the type sh.keptn.event.get-sli.triggered, to inform the respective data source services to retrieve the values for the SLIs defined in the slo.yaml file. It will be installed by default during installation of Keptn, if not installed then please follow steps in this link."

Not quite... You're correct that the lighthouse service responds to evaluation.triggered but internally it then sends a get-sli.triggered. The SLI provider tool is the one listening for get-sli.triggered.
When the SLI provider has retrieved metrics, the SLI provider sends get-sli.finished. The lighthouse service is also configured to listen for get-sli.finished so it starts working again and this time it does the evaluation.

- Typo: `Configuring a Data Soruce` > `Configuring a Data Source`

- Please use the `keptn configure monitoring` command (or they can craft and send via API) rather than only addding the config map. Why? Depending on the sli-provider author, the `keptn configure monitoring` command actually sends the `sh.keptn.event.monitoring.configure` CloudEvent.
This can then be used by one or more services to do different things (depending on what the author of the SLI provider service wants). For example, the Prometheus service configures Prom targets and alerts: https://github.com/keptn-contrib/prometheus-service/blob/9624777a0dbaab6afff7b4beb5e0ca8f94d82da2/eventhandling/configure_event_handler.go#L106

In short, just adding the ConfigMap might work, you might miss something (depending on what the sli provider author decides) and of course, what the author decides can change at any time so things might "just start breaking". The expectation is that authors respond to CloudEvents so that's what they're expecting.

**Comment by jainammm** [comment]
@agardnerIT 

# docs/k6-jes-example/README.md
1. Made the changes suggested.
2. Here for the K6 extension we need to enable remote-write... otherwise it throws this error
```time="2022-07-19T19:40:24Z" level=error msg="Failed to store timeseries." error="server returned HTTP status 404 Not Found: remote write receiver needs to be enabled with --enable-feature=remote-write-receiver"```
I can see that K6 uses remote write URL for writing while this project directly used ```from prometheus_client import CollectorRegistry, Counter, Gauge, push_to_gateway```. There might lie the difference. 

# docs/k6-prometheus-quality-gate-example/README.md
1. Made the changes
2. just adding `keptn configure monitoring` won't require the addition of ConfigMap file, right?
